cpp/src/traversal/opg/frontier_expansion.cu

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <raft/handle.hpp>
+#include "load_balance.cuh"
+#include "bfs_comms.cuh"
+
+namespace detail {
+
+namespace opg {
+
+struct bitwise_or {
+  __device__ unsigned operator()(unsigned& a, unsigned & b) { return a | b; }
+};
+
+struct remove_visited {
+  __device__ unsigned operator()(unsigned& visited, unsigned& output) {
+    //OUTPUT AND VISITED - common bits between output and visited
+    //OUTPUT AND (NOT (OUTPUT AND VISITED))
+    // - remove common bits between output and visited from output
+    return (output & (~( output & visited )));
+  }
+};
+
+template <typename VT>
+struct bitwise_atomic_or {
+  unsigned * output_frontier_;
+  VT * predecessors_;
+
+  bitwise_atomic_or(
+      unsigned * output_frontier,
+      VT * predecessors) :
+    output_frontier_(output_frontier),
+    predecessors_(predecessors) {}
+
+  __device__ void operator()(VT src, VT dst) {
+    unsigned active_bit = static_cast<unsigned>(1)<<(dst % BitsPWrd<unsigned>);
+    unsigned prev_word =
+      atomicOr(output_frontier_ + (dst/BitsPWrd<unsigned>), active_bit);
+    //If this thread activates the frontier bitmap for a destination
+    //then the source is the predecessor of that destination
+    if (prev_word & active_bit == 0) {
+      predecessors_[dst] = src;
+    }
+  }
+};
+
+struct is_not_equal {
+  unsigned cmp_;
+  unsigned * flag_;
+  is_not_equal(unsigned cmp, unsigned * flag) : cmp_(cmp), flag_(flag) {}
+  __device__ void operator()(unsigned& val) {
+    if (val != cmp_) {
+      *flag_ = 1;
+    }
+  }
+};
+
+template <typename VT, typename ET, typename WT>
+void bfs(raft::handle_t const &handle,
+    cugraph::experimental::GraphCSRView<VT, ET, WT>& graph,
+    VT *predecessors,
+    const VT start_vertex) {
+
+  //We need to keep track if a vertex is visited or its status
+  //This needs to be done for all the vertices in the global graph
+  size_t word_count = number_of_words(graph.number_of_vertices);
+  rmm::device_vector<unsigned> input_frontier(word_count);
+  rmm::device_vector<unsigned> output_frontier(word_count);
+  rmm::device_vector<unsigned> visited(word_count);
```

Comment:
This will not scale beyond a small number of GPUs. Memory footprint stays same with the increasing number of GPUs. This cannot process a graph unless number_of_vertices bits * 3 fits inside a single GPU.

Reply:
The word count is (number of vertices / 32) since every unsigned word takes care of 32 vertices. That amounts to approx 10 percent memory overhead.

Reply:
OK, so V/32 words(V/8 bytes), and if I am not mistaken, this does not decrease with P (graph.number_of_vertices is the total number of vertices, right?). This is 10% memory overhead (with a single GPU?) but if other memory allocations go down with more GPUs, this will eventually become the dominating factor.

Reply:
Is this PR targeting a small number of GPUs? (2 or 4?)

Reply:
>  this does not decrease with P (graph.number_of_vertices is the total number of vertices, right?).

No it doesn't.

>  if other memory allocations go down with more GPUs
I am not allocating more frontiers in this section if more GPUs are present. Each of these vectors only needs to exist once per GPU. Input and output frontier for the local frontier calculation and visited for the global accumulation of visited vertices.

> Is this PR targeting a small number of GPUs? (2 or 4?)
This PR is not targeting small number of GPUs. Ideally around 32/64.

Reply:
I see. Please let me know if you are currently just experimenting and I jumped to code review too soon. I'd better not complain too much about the code that will not be merged :-)